drivers/flash/flash_stm32h7x.c

```diff
@@ -0,0 +1,534 @@
+}
+
+int flash_stm32_block_erase_loop(const struct device *dev,
+				 unsigned int offset,
+				 unsigned int len)
+{
+	int i;
+	int rc = 0;
+
+	for (i = get_sector(offset); i <= get_sector(offset + len - 1); ++i) {
+		rc = erase_sector(dev, i);
+		if (rc < 0) {
+			break;
+		}
+	}
+	return rc;
+}
+
+static int wait_write_queue(const struct device *dev, off_t offset)
+{
+	int64_t timeout_time = k_uptime_get() + 100;
+	FLASH_TypeDef *regs = FLASH_STM32_REGS(dev);
+	uint32_t sector = get_sector(offset);
+	volatile uint32_t *sr;
+
+	if (sector < FLASH_SECTOR_TOTAL) {
+		sr = &regs->SR1;
+#ifdef DUAL_BANK
+	} else if (sector < 2 * FLASH_SECTOR_TOTAL) {
+		sr = &regs->SR2;
+#endif
+	} else {
+		return -EINVAL;
+	}
+
+	while (*sr & FLASH_SR_QW) {
+		if (k_uptime_get() > timeout_time) {
+			LOG_ERR("Timeout! val: %d", 100);
+			return -EIO;
+		}
+	}
+
+	return 0;
+}
+
+static int write_ndwords(const struct device *dev,
+			 off_t offset, const uint64_t *data,
+			 uint8_t n)
+{
+	volatile uint64_t *flash = (uint64_t *)(offset
+						+ CONFIG_FLASH_BASE_ADDRESS);
+	FLASH_TypeDef *regs = FLASH_STM32_REGS(dev);
+	volatile uint32_t *cr;
+	uint32_t tmp;
+	int rc;
+	int i;
+	uint32_t sector = get_sector(offset);
+
+	if (sector < FLASH_SECTOR_TOTAL) {
+		cr = &regs->CR1;
+#ifdef DUAL_BANK
+	} else if (sector < 2 * FLASH_SECTOR_TOTAL) {
+		cr = &regs->CR2;
+#endif
+	} else {
+		return -EINVAL;
+	}
+
+	/* if the control register is locked, do not fail silently */
+	if (*cr & FLASH_CR_LOCK) {
+		return -EIO;
+	}
+
+	/* Check that no Flash main memory operation is ongoing */
+	rc = flash_stm32h7_wait_flash_idle(dev);
+	if (rc < 0) {
+		return rc;
+	}
+
+	/* Check if 256 bits location is erased */
+	for (i = 0; i < n; ++i) {
+		if (flash[i] != 0xFFFFFFFFFFFFFFFFUL) {
+			return -EIO;
+		}
+	}
+
+	/* Set the PG bit */
+	*cr |= FLASH_CR_PG;
+
+	/* Flush the register write */
+	tmp = *cr;
+
+	/* Perform the data write operation at the desired memory address */
+	for (i = 0; i < n; ++i) {
+		flash[i] = data[i];
+		wait_write_queue(dev, offset);
+	}
+
+	/* Wait until the BSY bit is cleared */
+	rc = flash_stm32h7_wait_flash_idle(dev);
+
+	/* Clear the PG bit */
+	*cr &= (~FLASH_CR_PG);
+
+	return rc;
+}
+
+int flash_stm32_write_range(const struct device *dev, unsigned int offset,
+			    const void *data, unsigned int len)
+{
+	int rc = 0;
+	int i = 0;
+	int j;
+	const uint8_t ndwords = FLASH_NB_32BITWORD_IN_FLASHWORD / 2;
+	const uint8_t nbytes = FLASH_NB_32BITWORD_IN_FLASHWORD * 4;
+	uint8_t unaligned_datas[nbytes];
+	int suboffset = (offset % nbytes);
+
+	if (suboffset != 0) {
+		/* Offset is not aligned on FLASHWORD boundary */
+		offset -= suboffset;
+		memset(unaligned_datas, 0xff, sizeof(unaligned_datas));
+		for (i = 0; i < len && i < nbytes; ++i) {
+			unaligned_datas[suboffset + i] = ((uint8_t *)data)[i];
+		}
+		rc = write_ndwords(dev, offset,
+				   (const uint64_t *)unaligned_datas,
+				   ndwords);
+		i = 32;
+		offset += 32;
+	}
```

Comment:
I don't think this is something we'd like to do.
If write request offset is not aligned on flash blocks boundaries we should reject and warn user, as done in other drivers.